cli/src/commands/evolog.rs

```diff
@@ -151,14 +159,36 @@ pub(crate) fn cmd_evolog(
     if !args.no_graph {
         let mut raw_output = formatter.raw()?;
         let mut graph = get_graphlog(graph_style, raw_output.as_mut());
-        for commit in commits {
-            let edges = commit
-                .predecessor_ids()
+
+        let commit_dag: Vec<Result<GraphNode<Commit>, ()>> = commits
+            .iter()
+            .map(|c| {
+                let edges = c
+                    .predecessors()
+                    .map(|p| GraphEdge::direct(p.unwrap().clone()))
+                    .collect_vec();
+                Ok((c.clone(), edges))
+            })
+            .collect_vec();
+
+        let iter_nodes: Box<dyn Iterator<Item = _>> = if args.reversed {
```

Comment:
Boxed iterator is no longer needed since `reversed_graph()` returns a `Vec`. You can collect `Vec<GraphNode<Commit>>` first, and reverse it as needed.
```rust
if reversed {
    reverse_graph(commit_dag.into_iter().map(Result::<_, Infallible>::Ok)).unwrap()
...
```
(https://doc.rust-lang.org/stable/std/convert/enum.Infallible.html is preferred because the type system ensures that `Err` is never constructed. In recent version of Rust, we can use `let Ok(_) = ..` instead of `unwrap()`.)

Reply:
I didn't know about `Infallible`! Thanks for pointing it out. I touched up this code a bit more. I'm not convinced I used `Infallible` correctly, but I got rid of the boxed iterators, so it's at least simpler / less noisy in that way.